File: classes.CS/CGolosAPI.cs
using System;
using System.Collections;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Threading.Tasks;

namespace GolosAPI.CS
{
	class CGolosAPI : IDisposable
	{

		#region Enums
		private enum EType
		{
			RPC,
			WS
		};
		#endregion

		#region Variables
		private EType m_eType;
		private CJson m_oJson;
		private CWebsocket m_oSocket;
		#endregion

		#region Constructors
		public CGolosAPI(string strHostname, ushort nPort)
		{
			m_oJson = new CJson(strHostname, nPort, "/rpc");
			m_eType = EType.RPC;
		}

		public CGolosAPI(string strURI)
		{
			m_oSocket = new CWebsocket(strURI);
			m_eType = EType.WS;
		}
		#endregion

		#region Private methods
		private string SendRequest(string strMethod, ArrayList strParams = null)
		{
		if( m_eType == EType.RPC)
			{
				return m_oJson.SendRequest(strMethod, strParams);
			}
				else
			{
				using (Task<string> t = m_oSocket.SendRequest(strMethod, strParams))
				{
					t.Wait();
						return t.Result;
				}
			}
		}
		#endregion

		#region protected methods
		protected JObject call_api(string strMethod)
		{
			return (JObject)JsonConvert.DeserializeObject<Dictionary<string, JObject>>(SendRequest(strMethod))["result"];
		}

		protected JObject call_api(string strMethod, ArrayList arrParams)
		{ 
			return (JObject)JsonConvert.DeserializeObject<Dictionary<string, JObject>>(SendRequest(strMethod, arrParams))["result"];
		}

		protected JArray call_api_array(string strMethod, ArrayList arrParams)
		{
			return (JArray)JsonConvert.DeserializeObject<Dictionary<string, JArray>>(SendRequest(strMethod, arrParams))["result"];
		}

		protected JArray call_api_array(string strMethod)
		{ 
			return (JArray)JsonConvert.DeserializeObject<Dictionary<string, JArray>>(SendRequest(strMethod))["result"];
		}

		protected JValue call_api_value(string strMethod)
		{

			return (JValue)JsonConvert.DeserializeObject<Dictionary<string, JValue>>(SendRequest(strMethod))["result"];
		}

		protected JValue call_api_value(string strMethod, ArrayList arrParams)
		{
			return (JValue)JsonConvert.DeserializeObject<Dictionary<string, JValue>>(SendRequest(strMethod, arrParams))["result"];

		}

		protected JToken call_api_token(string strMethod, ArrayList arrParams)
		{
			return (JToken)JsonConvert.DeserializeObject<Dictionary<string, JToken>>(SendRequest(strMethod, arrParams))["result"];
		}

		protected JToken call_api_token(string strMethod)
		{
			return (JToken)JsonConvert.DeserializeObject<Dictionary<string, JToken>>(SendRequest(strMethod))["result"];
		}

		public void Dispose()
		{
			((IDisposable)m_oSocket).Dispose();
		}
		#endregion
	}
}

File: classes.VB/CGolosAPI.vb
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq
Public Class CGolosAPI
    Implements IDisposable

#Region "Enums"

    Public Enum EType
        [RPC]
        [WS]
    End Enum

#End Region

#Region "Variables"
    Private m_eType As EType
    Private m_oJson As CJson
    Private m_oSocket As CWebSocket
#End Region

#Region "Constructors"
    Sub New(strHostname As String, nPort As UShort)
        m_oJson = New CJson(strHostname, nPort, "/rpc")
        m_eType = EType.RPC
    End Sub

    Sub New(strURI As String)
        m_oSocket = New CWebSocket(strURI)
        m_eType = EType.WS
    End Sub

#End Region

#Region "Private methods"
    Private Function SendRequest(strMethod As String, Optional strParams As ArrayList = Nothing) As String

        If m_eType = EType.RPC Then
            Return m_oJson.SendRequest(strMethod, strParams)
        Else
            Using t As Task(Of String) = m_oSocket.SendRequest(strMethod, strParams)
                t.Wait()
                Return t.Result
            End Using
        End If
    End Function

#End Region

#Region "Protected methods"
    Protected Function call_api(strMethod As String) As JObject
        Return JsonConvert.DeserializeObject(SendRequest(strMethod)).Item("result")
    End Function

    Protected Function call_api(strMethod As String, arrParams As ArrayList) As JObject
        Return JsonConvert.DeserializeObject(SendRequest(strMethod, arrParams)).Item("result")
    End Function

    Protected Function call_api_array(strMethod As String, arrParams As ArrayList) As JArray
        Return JsonConvert.DeserializeObject(SendRequest(strMethod, arrParams)).Item("result")
    End Function

    Protected Function call_api_array(strMethod As String) As JArray
        Return JsonConvert.DeserializeObject(SendRequest(strMethod)).Item("result")
    End Function

    Protected Function call_api_value(strMethod As String) As JValue
        Return JsonConvert.DeserializeObject(SendRequest(strMethod)).Item("result")
    End Function

    Protected Function call_api_value(strMethod As String, arrParams As ArrayList) As JValue
        Return JsonConvert.DeserializeObject(SendRequest(strMethod, arrParams)).Item("result")
    End Function

    Protected Function call_api_token(strMethod As String, arrParams As ArrayList) As JToken
        Return JsonConvert.DeserializeObject(SendRequest(strMethod, arrParams)).Item("result")
    End Function

    Protected Function call_api_token(strMethod As String) As JToken
        Return JsonConvert.DeserializeObject(SendRequest(strMethod)).Item("result")
    End Function

#End Region

#Region "IDisposable Support"
    Public Sub Dispose() Implements IDisposable.Dispose
        If Not m_oSocket Is Nothing Then
            m_oSocket.Dispose()
        End If
    End Sub
#End Region

End Class
